tests/test_inheritance.cpp

```diff
@@ -97,4 +97,27 @@ test_initializer inheritance([](py::module &m) {
             py::isinstance<Unregistered>(l[6])
         );
     });
+
+    m.def("test_matching_holder_type", []() {
+        struct Base { };
+        struct Derived : Base { };
+
+        try {
+            {
+                py::module m;
+                py::class_<Base, std::shared_ptr<Base>>(m, "Base");
+                py::class_<Derived, Base>(m, "Derived")
+                    .def(py::init<>());
+            }
+            {
+                py::module m;
+                py::class_<Base>(m, "Base");
+                py::class_<Derived, std::shared_ptr<Derived>, Base>(m, "Derived")
+                    .def(py::init<>());
+            }
```

Comment:
Unless I'm missing something, the second test is going to be skipped completely because the first one throws. Consider separate functions.

Also, `.def(py::init<>())` is probably not needed for the test.